notify-server/notify_server/clients/publisher.py

```
@@ -29,7 +29,10 @@ async def _send_task(connection: Connection, queue: Queue) -> None:
     try:
         while True:
             entry: QueueEntry = await queue.get()
-            await connection.send_multipart(entry.to_frames())
+            try:
+                await connection.send_multipart(entry.to_frames())
+            except Exception:
+                log.exception(f"Failed to send {entry}")
```

Should we really be swallowing all exceptions here? What exceptions do we anticipate from `send_multipart()` and `to_frames()`? Can we be more narrow?

If we have to fail, can we fail somewhere more transparent than this background task? For example, if a user of this API can cause an exception by trying to send a badly-formatted payload, is there a way to raise that error *when they first give us the payload,* instead of here, when we happen to be getting around to sending it?

---

Thinking about this some more: I think exception handling being tough here is a symptom of some funky internal implementation details, which also happen to leak through and cause some funky semantics on `send()`.

This might sound a bit handwavey because I'm rushing out the door. And I'm only like 20% sure how ZMQ works because I haven't read the guide yet. But bear with me.

-----

Our `publisher.Publisher` has an internal `Queue`. This internal queue happens to be of unbounded length (I'm not sure if the unbounded length is intentional, but that's how we make it in `publisher.create()`).

`publisher.Publisher.send()` returns after it places a message on that internal queue. To me, this...seems like an unnecessarily weak primitive? Wouldn't we want it to return when it successfully *gives ZMQ the message*? In a sense, ZMQ is itself a big queue, right? Why, then, are we only using it indirectly, through a feeder queue?

Using an internal feeder queue:

* Hides certain failures until it's too late to handle them properly, as discussed above in my comment about exception handling.
* Circumvents natural backpressure. [Backpressure is a good thing (and hidden buffers can be a bad thing)](https://vorpus.org/blog/some-thoughts-on-asynchronous-api-design-in-a-post-asyncawait-world/#bug-1-backpressure). If, for some reason, ZMQ can't accept messages as quickly as we're submitting them, we should slow down until ZMQ is ready.
* Smells like the kind of thing that will hide deadlocking bugs. I'm vaguely reminded of some Go bugs that I've seen. Handwavey, I know, sorry—but from a high level, you can imagine how a hidden buffer adds nondeterminism to whether a `put()` or `get()` operation will block, and that nondeterminism causes bugs because you have to remember to consider it *every time you `put()` or `get()`*, and that's hard.
* Weakens our message ordering guarantees by a lot. We can no longer say anything about the relative ordering of events submitted by two publishers to a common topic. If we're fine with this, then, okay—but then we should formally document what delivery semantics we *can* rely on, and we'll have to be very careful to build users of our notification system around those guarantees. Because that's the kind of breach of intuition that makes distributed systems hard.

Separately, `publisher.Publisher.send_nowait()` returns immediately, and either submits the message or fails silently depending on whether there's room in the internal queue. Again, I don't think this is a usefully strong primitive. But also, since the queue is unbounded, I think this will never actually fail, and so it's equivalent to `publisher.Publisher.send()`?

-----

tl;dr I think this is the code telling us that the internal feeder queue is questionable, and we should listen to the code. But let me know if this are unclear and I'll try to be a bit more concrete when I have some more time.

---

I made it way more complicated than necessary due to ignorance. The goal was to make sending an event be as non-obtrusive as possible. But I didn't realize that zmq already does that. 

Thanks for thinking about this.

I will remove the queuing.